exchanges/binance/binance_websocket.go

```diff
@@ -503,130 +504,82 @@ func (b *Binance) UpdateLocalBuffer(wsdp *WebsocketDepthStream) (bool, error) {
 	return false, err
 }
 
-// generateSubscriptions generates the default subscription set
 func (b *Binance) generateSubscriptions() (subscription.List, error) {
-	var channels = make([]string, 0, len(b.Features.Subscriptions))
-	for i := range b.Features.Subscriptions {
-		name, err := channelName(b.Features.Subscriptions[i])
-		if err != nil {
-			return nil, err
-		}
-		channels = append(channels, name)
-	}
-	var subscriptions subscription.List
-	pairs, err := b.GetEnabledPairs(asset.Spot)
-	if err != nil {
-		return nil, err
-	}
-	for y := range pairs {
-		for z := range channels {
-			lp := pairs[y].Lower()
-			lp.Delimiter = ""
-			subscriptions = append(subscriptions, &subscription.Subscription{
-				Channel: lp.String() + "@" + channels[z],
-				Pairs:   currency.Pairs{pairs[y]},
-				Asset:   asset.Spot,
-			})
+	for _, s := range b.Features.Subscriptions {
+		if s.Asset == asset.Empty {
+			// Handle backwards compatibility with config without assets, all binance subs are spot
+			s.Asset = asset.Spot
 		}
 	}
-	return subscriptions, nil
+	return b.Features.Subscriptions.ExpandTemplates(b)
 }
 
-// channelName converts a Subscription Config into binance format channel suffix
-func channelName(s *subscription.Subscription) (string, error) {
-	name, ok := subscriptionNames[s.Channel]
-	if !ok {
-		return name, fmt.Errorf("%w: %s", stream.ErrSubscriptionNotSupported, s.Channel)
+var subTemplate *template.Template
+
+// GetSubscriptionTemplate returns a subscription channel template
+func (b *Binance) GetSubscriptionTemplate(_ *subscription.Subscription) (*template.Template, error) {
+	var err error
+	if subTemplate == nil {
+		subTemplate, err = template.New("subscriptions.tmpl").
+			Funcs(template.FuncMap{
+				"interval": formatChannelInterval,
+				"levels":   formatChannelLevels,
+				"fmt":      currency.EMPTYFORMAT.Format,
+			}).
+			Parse(subTplText)
+	}
+	return subTemplate, err
+}
+
+func formatChannelLevels(s *subscription.Subscription) string {
+	if s.Levels != 0 {
+		return strconv.Itoa(s.Levels)
 	}
+	return ""
+}
 
+func formatChannelInterval(s *subscription.Subscription) string {
 	switch s.Channel {
 	case subscription.OrderbookChannel:
-		if s.Levels != 0 {
-			name += "@" + strconv.Itoa(s.Levels)
-		}
 		if s.Interval.Duration() == time.Second {
-			name += "@1000ms"
-		} else {
-			name += "@" + s.Interval.Short()
+			return "@1000ms"
 		}
+		return "@" + s.Interval.Short()
 	case subscription.CandlesChannel:
-		name += "_" + s.Interval.Short()
+		return "_" + s.Interval.Short()
```

Comment:
I'm raising this for awareness given that it is preexisting. Config `Interval` is a `kline.Interval` 
->`time.Duration` and so it does not support days, weeks etc. One cannot put in `1d` into their config to get 1 day candles. If they put in `24h` they also don't get data because Binance supports `1d`

Reply:
Ack.